# <font color=green>__Part 1__</font>

In [1]:
import pandas as pd
import numpy as np
!pip3 install lxml

     |████████████████████████████████| 5.5 MB 3.2 MB/s eta 0:00:01


__We will use pandas' read_html function and then choose the first table, which contains the relevant information:__

In [2]:
dataframe = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]
dataframe.head(3)

Postal Code       Borough Neighbourhood
0         M1A  Not assigned  Not assigned
1         M2A  Not assigned  Not assigned
2         M3A    North York     Parkwoods

In [3]:
# verifying that the "Not assigned" entries are indeed of type string
type(dataframe.iloc[0,1])

str

__As per the assignment, we drop all the rows for which "Borough" is unassigned:__

In [4]:
dataframe = dataframe[dataframe.Borough != 'Not assigned'].reset_index(drop = 'True')

dataframe.head(3)

Postal Code           Borough              Neighbourhood
0         M3A        North York                  Parkwoods
1         M4A        North York           Victoria Village
2         M5A  Downtown Toronto  Regent Park, Harbourfront

__According to the assignment, there are duplicates in the 'Postal Code' column. It is, however, possible, that the wikipedia page has been updated after the assignment was posted. It turns out that each postal code is listed only once:__

In [5]:
dataframe.shape[0] - len(dataframe['Postal Code'].unique())

0

__Finally, we verify that there are no rows with unassigned neighbourhood:__

In [6]:
"Not assigned" in dataframe.Neighbourhood.values

False

__As per the assignment, we print the number of rows using the .shape method:__

In [7]:
print(dataframe.shape[0])

103


# <font color=green>__Part 2__</font>

__First we define a function that will get the coordinates associated to a certain postal code using the Geocoder package. As suggested in the assignment, we take measures to deal with the unreliability of the package:__

In [8]:
!pip3 install geocoder
import geocoder # import geocoder

def coords(postal_code):
    # initialize your variable to None
    lat_lng_coords = None

    # loop until you get the coordinates
    while(lat_lng_coords is None):
      g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
      lat_lng_coords = g.latlng

    return lat_lng_coords

     |████████████████████████████████| 98 kB 2.7 MB/s eta 0:00:011
     |████████████████████████████████| 829 kB 7.7 MB/s eta 0:00:01
  Created wheel for future: filename=future-0.18.2-py3-none-any.whl size=491058 sha256=726553e78066417ee2dfe42291bfe9aea2ffe2273848016a86028efe97615666
  Stored in directory: /home/jovyan/.cache/pip/wheels/56/b0/fe/4410d17b32f1f0c3cf54cdfb2bc04d7b4b8f4ae377e2229ba0
Successfully built future


__However, attempting to use the thusly defined function, we get an error message. Upon closer inspection it turns out that an API key is needed. Since I do not have an API key, I will use the csv file provided in the assignment instead:__

In [9]:
df_coord = pd.read_csv('https://cocl.us/Geospatial_data')
df_coord.head(3)

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711

In [10]:
# We verify that there are no duplicates in the 'Postal Code' column
print(df_coord.shape[0]-len(df_coord['Postal Code'].unique()))

# We verify that there are precisely as many postal codes as in the other dataframe:
df_coord.shape[0] - dataframe.shape[0]

0


0

__We now fuse these two dataframes together, noting that the respective 'Postal Code' columns are ordered in a different way:__

In [11]:
# We use 'Postal Code' as index for both dataframes. This is justified, since there are no duplicates (see above)
df_coord.set_index('Postal Code', inplace = True)
dataframe.set_index('Postal Code', inplace = True)

# Now we can combine the dataframes based on index, where the different order does not matter
dataframe = pd.merge(dataframe, df_coord, left_index=True, right_index=True)

# Now we just reset the index and inspect the first twelve rows of the resulting dataframe:
dataframe.reset_index(inplace = True)
dataframe.head(12)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   

                                        Neighbourhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

# <font color=green>__Part 3__</font>

In [12]:
!conda install -c conda-forge folium=0.5.0 --yes
import folium
import json
import requests
from pandas.io.json import json_normalize

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.8.2
  latest version: 4.8.4

Please update conda by running

    $ conda update -n base conda



## Package Plan ##

  environment location: /srv/conda/envs/notebook

  added / updated specs:
    - folium=0.5.0


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    folium-0.5.0               |             py_0          45 KB  conda-forge
    openssl-1.1.1g             |       h516909a_1         2.1 MB  conda-forge
    vincent-0.4.4              |             py_1          28 KB  conda-forge
    ------------------------------------------------------------
                                           Total:         2.2 MB

The following NEW packages will be INSTALLED:

  folium             conda-forge/noarch::folium

__We start by displaying the neighbourhoods on a map:__

In [13]:
map_Toronto = folium.Map(location=[43.6532, -79.3832], zoom_start=10)

for lat, lng, borough, neighborhood, postalcode in zip(dataframe['Latitude'], dataframe['Longitude'],
                                                       dataframe['Borough'], dataframe['Neighbourhood'],
                                                       dataframe['Postal Code']):
    label = folium.Popup('{}, {}, {}'.format(postalcode, neighborhood, borough), parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

__We will use Foursquare in order to obtain information on the venues in these neighbourhoods:__

In [14]:
CLIENT_ID = 'removed for privacy reasons' # your Foursquare ID
CLIENT_SECRET = 'removed for privacy reasons' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

__We borrow the following function from one of the labs in the course:__

In [15]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

__We now go through the postal codes of Toronto and use the Foursquare API to find all the venues within radius=500 of the coordinates associated to the given postal codes; we limit ourselves to the top 100 venues for each postal code:__

In [16]:
postalcodes = dataframe['Postal Code']
neighbourhoods = dataframe['Neighbourhood']
latitudes = dataframe['Latitude']
longitudes = dataframe['Longitude']

venues_list=[]
for p_code, nhbd, lat, lng in zip(postalcodes, neighbourhoods, latitudes, longitudes):
            
    # create the API request URL
    url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
        CLIENT_ID, 
        CLIENT_SECRET, 
        VERSION, 
        lat, 
        lng, 
        500, 
        100)
            
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
        
    # return only relevant information for each nearby venue
    venues_list.append([(
        p_code, 
        nhbd, 
        lat, 
        lng, 
        v['venue']['name'], 
        v['venue']['location']['lat'], 
        v['venue']['location']['lng'],  
        v['venue']['categories'][0]['name']) for v in results])

nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
nearby_venues.columns = ['Postal Code', 
                  'Neighbourhood',
                  'Latitude', 
                  'Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    


__We briefly inspect the dataframe we just created:__

In [17]:
nearby_venues.head(20)

Postal Code              Neighbourhood   Latitude  Longitude  \
0          M3A                  Parkwoods  43.753259 -79.329656   
1          M3A                  Parkwoods  43.753259 -79.329656   
2          M4A           Victoria Village  43.725882 -79.315572   
3          M4A           Victoria Village  43.725882 -79.315572   
4          M4A           Victoria Village  43.725882 -79.315572   
5          M4A           Victoria Village  43.725882 -79.315572   
6          M4A           Victoria Village  43.725882 -79.315572   
7          M4A           Victoria Village  43.725882 -79.315572   
8          M5A  Regent Park, Harbourfront  43.654260 -79.360636   
9          M5A  Regent Park, Harbourfront  43.654260 -79.360636   
10         M5A  Regent Park, Harbourfront  43.654260 -79.360636   
11         M5A  Regent Park, Harbourfront  43.654260 -79.360636   
12         M5A  Regent Park, Harbourfront  43.654260 -79.360636   
13         M5A  Regent Park, Harbourfront  43.654260 -79.360636   
14         M5A  Regent Park, Harbourfront  43.654260 -79.360636   
15         M5A  Regent Park, Harbourfront  43.654260 -79.360636   
16         M5A  Regent Park, Harbourfront  43.654260 -79.360636   
17         M5A  Regent Park, Harbourfront  43.654260 -79.360636   
18         M5A  Regent Park, Harbourfront  43.654260 -79.360636   
19         M5A  Regent Park, Harbourfront  43.654260 -79.360636   

                                        Venue  Venue Latitude  \
0                             Brookbanks Park       43.751976   
1                               Variety Store       43.751974   
2                      Victoria Village Arena       43.723481   
3                                 Tim Hortons       43.725517   
4                                   Portugril       43.725819   
5                                    The Frig       43.727051   
6   Eglinton Ave E & Sloane Ave/Bermondsey Rd       43.726086   
7                                  Pizza Nova       43.725824   
8                            Roselle Desserts       43.653447   
9                               Tandem Coffee       43.653559   
10                     Cooper Koo Family YMCA       43.653249   
11                        Body Blitz Spa East       43.654735   
12                             Impact Kitchen       43.656369   
13                            Corktown Common       43.655618   
14           The Distillery Historic District       43.650244   
15                         Morning Glory Cafe       43.653947   
16                         The Extension Room       43.653313   
17                   Dominion Pub and Kitchen       43.656919   
18                   Distillery Sunday Market       43.650075   
19                        SOMA chocolatemaker       43.650622   

    Venue Longitude         Venue Category  
0        -79.332140                   Park  
1        -79.333114      Food & Drink Shop  
2        -79.315635           Hockey Arena  
3        -79.313103            Coffee Shop  
4        -79.312785  Portuguese Restaurant  
5        -79.317418      French Restaurant  
6        -79.313620           Intersection  
7        -79.312860            Pizza Place  
8        -79.362017                 Bakery  
9        -79.361809            Coffee Shop  
10       -79.358008    Distribution Center  
11       -79.359874                    Spa  
12       -79.356980             Restaurant  
13       -79.356211                   Park  
14       -79.359323          Historic Site  
15       -79.361149         Breakfast Spot  
16       -79.359725   Gym / Fitness Center  
17       -79.358967                    Pub  
18       -79.361832         Farmers Market  
19       -79.358127         Chocolate Shop

__It is important to note that a venue from this dataframe does not necessarily have the postal code from the 'Postal Code' column and it does not necessarily lie in (any of) the neighbourhood(s) from the Neighbourhood column; this is due to the fact that we were just looking for venues within a certain radius of the coordinates we specified, without any further requirements. Also note that it is possible that several venues were listed more than once; this will not matter to us, since we will just see them as being accessible from all the postal codes they are listed for.__

__We now inspect the size of this dataframe and see how many venues can be found for each of the postal codes:__

In [18]:
print(nearby_venues.shape)
nearby_venues[['Postal Code', 'Venue']].groupby('Postal Code').count().rename(columns={"Venue": "Number of Venues"})

(2139, 8)


Number of Venues
Postal Code                  
M1B                         1
M1C                         1
M1E                         8
M1G                         3
M1H                         9
M1J                         2
M1K                         6
M1L                        10
M1M                         2
M1N                         4
M1P                         7
M1R                         4
M1S                         4
M1T                        12
M1V                         3
M1W                        13
M2H                         4
M2J                        68
M2K                         4
M2N                        33
M2P                         3
M2R                         6
M3A                         2
M3B                         4
M3C                        23
M3H                        22
M3J                         7
M3K                         2
M3L                         5
M3M                         3
...                       ...
M6A                        11
M6B                         4
M6C                         3
M6E                         4
M6G                        17
M6H                        16
M6J                        44
M6K                        22
M6L                         4
M6M                         4
M6N                         4
M6P                        25
M6R                        14
M6S                        35
M7A                        34
M7R                        12
M7Y                        18
M8V                        14
M8W                         8
M8X                         4
M8Y                         3
M8Z                        14
M9C                         7
M9L                         1
M9M                         2
M9N                         2
M9P                         8
M9R                         4
M9V                         8
M9W                         4

[97 rows x 1 columns]

__Since there are fewer than 103 rows, we see that no venues were found for several of the neighbourhoods.__

__We view the number of different venue categories:__

In [19]:
len(nearby_venues['Venue Category'].unique())

269

__Now we inspect how many venues there are in each category for every postal code:__

In [20]:
venues_per_category = pd.get_dummies(nearby_venues[['Venue Category']], prefix="", prefix_sep="")

# add postal code column back to dataframe
venues_per_category['Postal Code'] = nearby_venues['Postal Code']  

# now we group the rows together by postal code and sum up
venues_per_category = venues_per_category.groupby('Postal Code').sum()
venues_per_category
 

Accessories Store  Afghan Restaurant  Airport  \
Postal Code                                                  
M1B                          0                  0        0   
M1C                          0                  0        0   
M1E                          0                  0        0   
M1G                          0                  0        0   
M1H                          0                  0        0   
M1J                          0                  0        0   
M1K                          0                  0        0   
M1L                          0                  0        0   
M1M                          0                  0        0   
M1N                          0                  0        0   
M1P                          0                  0        0   
M1R                          0                  0        0   
M1S                          0                  0        0   
M1T                          0                  0        0   
M1V                          0                  0        0   
M1W                          0                  0        0   
M2H                          0                  0        0   
M2J                          0                  0        0   
M2K                          0                  0        0   
M2N                          0                  0        0   
M2P                          0                  0        0   
M2R                          0                  0        0   
M3A                          0                  0        0   
M3B                          0                  0        0   
M3C                          0                  0        0   
M3H                          0                  0        0   
M3J                          0                  0        0   
M3K                          0                  0        1   
M3L                          0                  0        0   
M3M                          0                  0        0   
...                        ...                ...      ...   
M6A                          1                  0        0   
M6B                          0                  0        0   
M6C                          0                  0        0   
M6E                          0                  0        0   
M6G                          0                  0        0   
M6H                          0                  0        0   
M6J                          0                  0        0   
M6K                          0                  0        0   
M6L                          0                  0        0   
M6M                          0                  0        0   
M6N                          0                  0        0   
M6P                          0                  0        0   
M6R                          0                  0        0   
M6S                          0                  0        0   
M7A                          0                  0        0   
M7R                          0                  0        0   
M7Y                          0                  0        0   
M8V                          0                  0        0   
M8W                          0                  0        0   
M8X                          0                  0        0   
M8Y                          0                  0        0   
M8Z                          0                  0        0   
M9C                          0                  0        0   
M9L                          0                  0        0   
M9M                          0                  0        0   
M9N                          0                  0        0   
M9P                          0                  0        0   
M9R                          0                  0        0   
M9V                          0                  0        0   
M9W                          0                  0        0   

             Airport Food Court  Airport Gate  Airport Lounge  \
Postal Code                              

__In order to recover the most common venues for each postal code, we borrow the following function from one of the labs:__

In [21]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

__We now create a dataframe which contains the 5 most common venues for each postal code (assuming that at least 5 exist, note the default order otherwise):__

In [22]:
neighborhoods_venues_sorted = pd.DataFrame(columns = ['Postal Code', '1st Most Common Venue',
                                                      '2nd Most Common Venue', '3rd Most Common Venue',
                                                      '4th Most Common Venue', '5th Most Common Venue'])
neighborhoods_venues_sorted['Postal Code'] = venues_per_category.reset_index()['Postal Code']

for ind in np.arange(venues_per_category.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(venues_per_category.iloc[ind, :], 5)

neighborhoods_venues_sorted

Postal Code          1st Most Common Venue   2nd Most Common Venue  \
0          M1B           Fast Food Restaurant              Donut Shop   
1          M1C                            Bar             Yoga Studio   
2          M1E            Rental Car Location          Medical Center   
3          M1G                    Coffee Shop       Korean Restaurant   
4          M1H                           Bank                  Bakery   
5          M1J                    Pizza Place              Playground   
6          M1K               Department Store          Discount Store   
7          M1L                         Bakery                Bus Line   
8          M1M                          Motel     American Restaurant   
9          M1N                College Stadium            Skating Rink   
10         M1P              Indian Restaurant               Pet Store   
11         M1R                     Smoke Shop                  Bakery   
12         M1S                         Lounge            Skating Rink   
13         M1T                    Pizza Place                    Bank   
14         M1V                           Park        Asian Restaurant   
15         M1W           Fast Food Restaurant      Chinese Restaurant   
16         M2H                           Pool             Golf Course   
17         M2J                 Clothing Store             Coffee Shop   
18         M2K                           Café                    Bank   
19         M2N               Ramen Restaurant                    Café   
20         M2P                           Park       Convenience Store   
21         M2R                           Bank           Grocery Store   
22         M3A                           Park       Food & Drink Shop   
23         M3B                            Gym    Caribbean Restaurant   
24         M3C                            Gym             Coffee Shop   
25         M3H                    Coffee Shop                    Bank   
26         M3J         Furniture / Home Store           Metro Station   
27         M3K                           Park                 Airport   
28         M3L                           Park           Grocery Store   
29         M3M                     Food Truck        Business Service   
..         ...                            ...                     ...   
67         M6A                 Clothing Store  Furniture / Home Store   
68         M6B            Japanese Restaurant        Sushi Restaurant   
69         M6C                          Field                   Trail   
70         M6E                           Park           Women's Store   
71         M6G                  Grocery Store                    Café   
72         M6H                         Bakery                Pharmacy   
73         M6J                            Bar             Coffee Shop   
74         M6K                           Café          Breakfast Spot   
75         M6L                           Park                  Bakery   
76         M6M                 Discount Store                     Bar   
77         M6N              Convenience Store           Grocery Store   
78         M6P             Mexican Restaurant         Thai Restaurant   
79         M6R                 Breakfast Spot               Gift Shop   
80         M6S                    Coffee Shop                    Café   
81         M7A                    Coffee Shop       College Cafeteria   
82         M7R                          Hotel             Coffee Shop   
83         M7Y                    Yoga Studio           Auto Workshop   
84         M8V                            Gym              Hobby Shop   
85         M8W                    Pizza Place                     Gym   
86         M8X                           Park                    Pool   
87         M8Y                           Pool          Baseball Field   
88         M8Z                  Grocery Store       Convenience Store   
89         M9C                   Liquor Store               P

__We will now cluster the neighbourhoods. In contrast to the labs, we did not compute the frequency of the venue categories, but rather the total number of venues in a given category. Hence it would not be surprising if the clustering done here would cluster neighbourhoods together based on total numbers of venues rather than based on frequency of occurence, especially if the number of clusters is chosen to be small.__

__Because of this, I expect there to be several clusters which are central, and only one or two clusters, in which neighbourhoods with greater distance to the center are grouped together.__

In [23]:
from sklearn.cluster import KMeans

In [24]:
venues_clustering = venues_per_category.reset_index().drop('Postal Code', axis = 1)

kmeans = KMeans(n_clusters = 5, random_state=0).fit(venues_clustering)

__We create a new dataframe, to be used for visualization purposes:__

In [25]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

df_merged = dataframe

df_merged = df_merged.join(neighborhoods_venues_sorted.set_index('Postal Code'), on='Postal Code')

# for the neighbourhoods for which no venues were found there are some NaN entries in the corresponding rows
# we drop these rows

df_merged.dropna(axis = 0, inplace = True)
df_merged.head(10)

Postal Code           Borough                                Neighbourhood  \
0          M3A        North York                                    Parkwoods   
1          M4A        North York                             Victoria Village   
2          M5A  Downtown Toronto                    Regent Park, Harbourfront   
3          M6A        North York             Lawrence Manor, Lawrence Heights   
4          M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
6          M1B       Scarborough                               Malvern, Rouge   
7          M3B        North York                                    Don Mills   
8          M4B         East York              Parkview Hill, Woodbine Gardens   
9          M5B  Downtown Toronto                     Garden District, Ryerson   
10         M6B        North York                                    Glencairn   

     Latitude  Longitude  Cluster Labels  1st Most Common Venue  \
0   43.753259 -79.329656             1.0                   Park   
1   43.725882 -79.315572             1.0  Portuguese Restaurant   
2   43.654260 -79.360636             4.0            Coffee Shop   
3   43.718518 -79.464763             1.0         Clothing Store   
4   43.662301 -79.389494             4.0            Coffee Shop   
6   43.806686 -79.194353             1.0   Fast Food Restaurant   
7   43.745906 -79.352188             1.0                    Gym   
8   43.706397 -79.309937             1.0            Pizza Place   
9   43.657162 -79.378937             3.0         Clothing Store   
10  43.709577 -79.445073             1.0    Japanese Restaurant   

     2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0        Food & Drink Shop            Event Space  Ethiopian Restaurant   
1             Intersection      French Restaurant           Coffee Shop   
2                   Bakery                   Café                   Pub   
3   Furniture / Home Store  Vietnamese Restaurant           Coffee Shop   
4        College Cafeteria                    Gym            Hobby Shop   
6               Donut Shop     Dim Sum Restaurant                 Diner   
7     Caribbean Restaurant    Japanese Restaurant                  Café   
8                     Bank   Gym / Fitness Center          Intersection   
9              Coffee Shop                   Café        Cosmetics Shop   
10        Sushi Restaurant                    Pub         Metro Station   

   5th Most Common Venue  
0      Electronics Store  
1            Pizza Place  
2                   Park  
3            Event Space  
4               Beer Bar  
6         Discount Store  
7            Yoga Studio  
8                   Café  
9    Japanese Restaurant  
10          Dessert Shop

__We now visualize the clusters using Folium:__

In [26]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [27]:
map_clusters = folium.Map(location = [43.6532, -79.3832], zoom_start = 10)

x = np.arange(5)
ys = [i + x + (i*x)**2 for i in range(5)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(df_merged['Latitude'], df_merged['Longitude'], df_merged['Postal Code'],
                                  df_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

__One observes that our intuition was correct: distance to center seems to be one of the distinguishing features between clusters.__

__Of course "distance to center" is just one out of many aspects to consider. Let's have a closer look at the clusters:__

In [28]:
df_merged.loc[df_merged['Cluster Labels'] == 0, df_merged.columns[[1] + list(range(5, df_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
25  Downtown Toronto             0.0         Grocery Store   
37      West Toronto             0.0                   Bar   
41      East Toronto             0.0      Greek Restaurant   
43      West Toronto             0.0                  Café   
54      East Toronto             0.0                  Café   
55        North York             0.0        Sandwich Place   
59        North York             0.0      Ramen Restaurant   
74   Central Toronto             0.0                  Café   
79   Central Toronto             0.0        Sandwich Place   
80  Downtown Toronto             0.0                  Café   
81      West Toronto             0.0           Coffee Shop   
96  Downtown Toronto             0.0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
25                  Café                  Park           Coffee Shop   
37           Coffee Shop      Asian Restaurant                  Café   
41           Coffee Shop    Italian Restaurant            Restaurant   
43        Breakfast Spot           Coffee Shop         Burrito Place   
54           Coffee Shop                Bakery               Brewery   
55            Restaurant    Italian Restaurant           Coffee Shop   
59                  Café        Sandwich Place           Coffee Shop   
74        Sandwich Place           Coffee Shop          Liquor Store   
79          Dessert Shop           Pizza Place      Sushi Restaurant   
80                Bakery                   Bar   Japanese Restaurant   
81                  Café    Italian Restaurant           Pizza Place   
96    Chinese Restaurant    Italian Restaurant                  Park   

      5th Most Common Venue  
25       Italian Restaurant  
37    Vietnamese Restaurant  
41   Furniture / Home Store  
43               Restaurant  
54      American Restaurant  
55  Comfort Food Restaurant  
59              Pizza Place  
74        Indian Restaurant  
79                      Gym  
80                Bookstore  
81                      Pub  
96              Pizza Place

In [29]:
df_merged.loc[df_merged['Cluster Labels'] == 1, df_merged.columns[[1] + list(range(5, df_merged.shape[1]))]]

Borough  Cluster Labels   1st Most Common Venue  \
0          North York             1.0                    Park   
1          North York             1.0   Portuguese Restaurant   
3          North York             1.0          Clothing Store   
6         Scarborough             1.0    Fast Food Restaurant   
7          North York             1.0                     Gym   
8           East York             1.0             Pizza Place   
10         North York             1.0     Japanese Restaurant   
12        Scarborough             1.0                     Bar   
13         North York             1.0                     Gym   
14          East York             1.0                    Park   
16               York             1.0                   Field   
17          Etobicoke             1.0            Liquor Store   
18        Scarborough             1.0     Rental Car Location   
19       East Toronto             1.0                   Trail   
21               York             1.0                    Park   
22        Scarborough             1.0             Coffee Shop   
23          East York             1.0             Coffee Shop   
26        Scarborough             1.0                    Bank   
27         North York             1.0                    Pool   
28         North York             1.0             Coffee Shop   
29          East York             1.0          Sandwich Place   
31       West Toronto             1.0                  Bakery   
32        Scarborough             1.0             Pizza Place   
34         North York             1.0  Furniture / Home Store   
35          East York             1.0                    Park   
38        Scarborough             1.0        Department Store   
39         North York             1.0                    Café   
40         North York             1.0                    Park   
44        Scarborough             1.0                  Bakery   
46         North York             1.0                    Park   
..                ...             ...                     ...   
62    Central Toronto             1.0                  Garden   
63               York             1.0       Convenience Store   
64               York             1.0                    Park   
65        Scarborough             1.0       Indian Restaurant   
66         North York             1.0                    Park   
67    Central Toronto             1.0                    Park   
68    Central Toronto             1.0                   Trail   
69       West Toronto             1.0      Mexican Restaurant   
70          Etobicoke             1.0             Pizza Place   
71        Scarborough             1.0              Smoke Shop   
72         North York             1.0                    Bank   
73    Central Toronto             1.0             Coffee Shop   
75       West Toronto             1.0          Breakfast Spot   
76        Mississauga             1.0                   Hotel   
77          Etobicoke             1.0                    Park   
78        Scarborough             1.0                  Lounge   
82        Scarborough             1.0             Pizza Place   
85        Scarborough             1.0                    Park   
86    Central Toronto             1.0             Coffee Shop   
87   Downtown Toronto             1.0          Airport Lounge   
88          Etobicoke             1.0                     Gym   
89          Etobicoke             1.0           Grocery Store   
90        Scarborough             1.0    Fast Food Restaurant   
91   Downtown Toronto             1.0                    Park   
93          Etobicoke             1.0             Pizza Place   
94          Etobicoke             1.0     Rental Car Location   
98          Etobicoke             1.0                    Park   
100      East Toronto             1.0             Yoga Studio   
101         Etobicoke             1.0                    Pool   
102         Etobicoke             1.0           G

In [30]:
df_merged.loc[df_merged['Cluster Labels'] == 2, df_merged.columns[[1] + list(range(5, df_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
30  Downtown Toronto             2.0           Coffee Shop   
36  Downtown Toronto             2.0           Coffee Shop   
42  Downtown Toronto             2.0           Coffee Shop   
48  Downtown Toronto             2.0           Coffee Shop   
97  Downtown Toronto             2.0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
30                  Café            Restaurant                 Hotel   
36              Aquarium                 Hotel                  Café   
42                 Hotel                  Café            Restaurant   
48            Restaurant                  Café                 Hotel   
97                  Café                 Hotel            Restaurant   

   5th Most Common Venue  
30                   Gym  
36            Restaurant  
42           Salad Place  
48                   Gym  
97                   Gym

In [31]:
df_merged.loc[df_merged['Cluster Labels'] == 3, df_merged.columns[[1] + list(range(5, df_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
9   Downtown Toronto             3.0        Clothing Store   
33        North York             3.0        Clothing Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9            Coffee Shop                  Café        Cosmetics Shop   
33           Coffee Shop  Fast Food Restaurant            Restaurant   

   5th Most Common Venue  
9    Japanese Restaurant  
33         Women's Store

In [32]:
df_merged.loc[df_merged['Cluster Labels'] == 4, df_merged.columns[[1] + list(range(5, df_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2   Downtown Toronto             4.0           Coffee Shop   
4   Downtown Toronto             4.0           Coffee Shop   
15  Downtown Toronto             4.0           Coffee Shop   
20  Downtown Toronto             4.0           Coffee Shop   
24  Downtown Toronto             4.0           Coffee Shop   
84  Downtown Toronto             4.0                  Café   
92  Downtown Toronto             4.0           Coffee Shop   
99  Downtown Toronto             4.0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
2                 Bakery                  Café                    Pub   
4      College Cafeteria                   Gym             Hobby Shop   
15                  Café          Cocktail Bar    American Restaurant   
20                Bakery          Cocktail Bar            Cheese Shop   
24        Sandwich Place                  Café     Italian Restaurant   
84           Coffee Shop    Mexican Restaurant  Vietnamese Restaurant   
92                  Café            Restaurant               Beer Bar   
99   Japanese Restaurant      Sushi Restaurant                Gay Bar   

            5th Most Common Venue  
2                            Park  
4                        Beer Bar  
15                     Restaurant  
20                           Café  
24            Japanese Restaurant  
84  Vegetarian / Vegan Restaurant  
92            Japanese Restaurant  
99                     Restaurant

__Cluster 0 contains neighbourhoods which are not central. The neighbourhoods in said cluster have a lot of coffee shops, cafés and restaurants.__

__Cluster 1 is the largest cluster. Its distinguishing property is that the neighbourhoods contained in it are relatively far from the center and have a low total number of venues. Since these neighbourhoods are non-central, it is not surprising that many of them have parks as their most common venue.__

__Cluster 2 is one of two clusters consisting of exclusively central neighbourhoods, the other one being Cluster 4. Non-surprisingly - given the central location- the most common venue for the neighbourhoods in either cluster is "Coffee Shop". Cluster 2 seems to be more of a hotel area, whereas Cluster 4 seems to be more of a restaurant area. However, there is not enough evidence to draw that conclusion with any reasonable amount of certainty.__

__Finally, Cluster 3 is the smallest cluster with only two neighbourhoods, one of them being central, the other one being quite far removed from the center. The neighbourhoods in this cluster seem to be connected to clothing and fashion.__